runners/spark/src/main/java/org/apache/beam/runners/spark/translation/EvaluationContext.java

```diff
@@ -126,13 +136,29 @@ public void putDataset(PValue pvalue, Dataset dataset) {
     } catch (IllegalStateException e) {
       // name not set, ignore
     }
+    if (shouldCache(pvalue)) {
+      dataset.cache(storageLevel());
+    }
     datasets.put(pvalue, dataset);
     leaves.add(dataset);
   }
 
   <T> void putBoundedDatasetFromValues(
       PTransform<?, ? extends PValue> transform, Iterable<T> values, Coder<T> coder) {
-    datasets.put(getOutput(transform), new BoundedDataset<>(values, jsc, coder));
+    PValue output = getOutput(transform);
+    if (shouldCache(output)) {
+      // eagerly create the RDD, as it will be reused.
+      Iterable<WindowedValue<T>> elems = Iterables.transform(values,
+          WindowingHelpers.<T>windowValueFunction());
+      WindowedValue.ValueOnlyWindowedValueCoder<T> windowCoder =
+          WindowedValue.getValueOnlyCoder(coder);
+      JavaRDD<WindowedValue<T>> rdd =
+          getSparkContext().parallelize(CoderHelpers.toByteArrays(elems, windowCoder))
+          .map(CoderHelpers.fromByteFunction(windowCoder));
+      putDataset(transform, new BoundedDataset<>(rdd));
+    } else {
+      datasets.put(getOutput(transform), new BoundedDataset<>(values, jsc, coder));
```

Comment:
Since we don't always create from Iterable lazy, it's worth a note too: `create a BoundedDataset that would create an RDD on demand`.